File: src/nasa/process-pioneer10-jupiter-trajectory-1.R
# Process NASA Pioneer 10 Jupiter Trajectory data
#

# Dependencies
library(tidyverse)
library(lubridate)
library(data.table) # For fread
library(sphereplot)
library(rgl)

# Download Data

# Download Trajectory data
# Format information see
# https://spdf.gsfc.nasa.gov/pub/data/pioneer/pioneer10/traj/jupiter/p10trjjup_fmt.txt
rtraj <- fread("https://spdf.gsfc.nasa.gov/pub/data/pioneer/pioneer10/traj/jupiter/p10trjjup.asc")

# Convert date time
traj <- rtraj %>% mutate(ts = date_decimal(V1) + as.duration(86400 * V2))

# Select and rename
traj <- traj %>% select(ts, srange = V3, seclat = V4, seclon = V5, prange = V6, peqlat = V7, peqlon = V8)

# Transforms 3D spherical coordinates to Cartesian coordinates.
scart <- sph2car(traj$seclon, traj$seclat, traj$srange)
pcart <- sph2car(traj$peqlon, traj$peqlat, traj$prange)

# Plot both trajectory
plot3d(scart)
plot3d(pcart)

File: src/python/nasa/process-pioneer10-jupiter-trajectory-1.py
# Process NASA Pioneer 10 Jupiter Trajectory data
#

# Dependencies
from matplotlib.pyplot import axis
import pandas as pd
import datetime as dt
import numpy as np

# Download Data

# Download Trajectory data
# Format information see
# https://spdf.gsfc.nasa.gov/pub/data/pioneer/pioneer10/traj/jupiter/p10trjjup_fmt.txt
rtraj_path = "https://spdf.gsfc.nasa.gov/pub/data/pioneer/pioneer10/traj/jupiter/p10trjjup.asc"
col_names = ['year', 'fdoy', 'srange', 'seclat', 'seclon', 'prange', 'peqlat', 'peqlon']
rtraj = pd.read_csv(rtraj_path, delim_whitespace=True, names=col_names, dtype=float)

# Fix date
rtraj['ts'] = rtraj.apply(lambda row: dt.date(int(row['year']), 1, 1) + dt.timedelta(seconds=86400 * row['fdoy']), axis=1)

print(rtraj)
